proof_test.go

```diff
@@ -0,0 +1,120 @@
+package merkletree
+
+import (
+	"bytes"
+	"testing"
+
+	"github.com/coniks-sys/libmerkleprefixtree-go/crypto"
+	"github.com/coniks-sys/libmerkleprefixtree-go/internal"
+)
+
+func computeLeafHash(ap *AuthenticationPath) (leafHash []byte) {
+	leaf := ap.Leaf()
+	if leaf.Commitment() != nil {
+		// user leaf node
+		leafHash = crypto.Digest(
+			[]byte{LeafIdentifier},              // K_leaf
+			[]byte(ap.TreeNonce()),              // K_n
+			[]byte(ap.Index()),                  // i
+			[]byte(util.IntToBytes(ap.Level())), // l
+			[]byte(leaf.Commitment()),           // commit(key|| value)
+		)
+	} else {
+		// empty leaf node
+		leafHash = crypto.Digest(
+			[]byte{EmptyBranchIdentifier},       // K_empty
+			[]byte(ap.TreeNonce()),              // K_n
+			[]byte(ap.Index()),                  // i
+			[]byte(util.IntToBytes(ap.Level())), // l
+		)
+	}
+	return
+}
+
+func computeProofHash(ap *AuthenticationPath, depth int) (leftHash, rightHash []byte) {
```

Comment:
I agree with @masomel here that starting from the deepest level is more intuitive.  Here's a little implementation of what that looks like, 780d27bcacdc9dc8d560dd3654c98fe68089547c. @c633 how do you feel about it?